metagraph/src/annotation/row_diff_builder.cpp

```diff
@@ -77,60 +78,54 @@ void build_successor(const std::string &graph_fname,
+    const uint64_t batch_size = 10'000'000;
+    // traverse BOSS table in parallel processing blocks of size |batch_size|
+    for (uint64_t start = 1; start <= graph.num_nodes(); start += batch_size) {
         std::vector<std::vector<uint64_t>> pred_buf(num_threads);
         std::vector<std::vector<uint64_t>> succ_buf(num_threads);
         std::vector<std::vector<bool>> pred_boundary_buf(num_threads);
 
-        #pragma omp parallel for num_threads(num_threads) schedule(static, 1)
-        for (uint32_t chunk = 0; chunk < std::max((uint32_t)1, num_threads); ++chunk) {
-            uint64_t start = batch * batch_size + chunk * chunk_size + 1;
-            for (uint64_t i = start;
-                 i < std::min(graph.num_nodes() + 1, start + chunk_size); ++i) {
-                uint64_t boss_idx = graph.kmer_to_boss_index(i);
-                if (dummy[boss_idx] || terminal[boss_idx]) {
-                    succ_buf[chunk].push_back(0);
-                } else {
-                    const graph::boss::BOSS::TAlphabet w
-                            = boss.get_W(boss_idx) % boss.alph_size;
-                    uint64_t next = w ? graph.boss_to_kmer_index(boss.fwd(boss_idx, w)) : 0;
-                    succ_buf[chunk].push_back(next);
-                }
-                uint64_t back_idx = boss.bwd(boss_idx);
-                boss.call_incoming_to_target(
-                        back_idx, boss.get_W(back_idx), [&](BOSS::edge_index incoming_edge) {
-                          // terminal and dummy predecessors are ignored; also ignore
-                          // predecessors for which boss_idx is not the last outgoing
-                          // edge (bc. we only traverse the last outgoing at a bifurcation)
-                          if (terminal[incoming_edge] || dummy[incoming_edge]
-                                  || boss.fwd(incoming_edge,
-                                              boss.get_W(incoming_edge) % boss.alph_size)
-                                          != boss_idx)
-                              return;
-
-                          pred_buf[chunk].push_back(graph.boss_to_kmer_index(incoming_edge));
-                          pred_boundary_buf[chunk].push_back(0);
-                        });
-                pred_boundary_buf[chunk].push_back(1);
+        // use static scheduling to make threads process ordered contiguous blocks
+        #pragma omp parallel for num_threads(num_threads) schedule(static)
+        for (uint64_t i = start; i < std::min(start + batch_size, graph.num_nodes() + 1); ++i) {
+            const size_t r = omp_get_thread_num();
+            uint64_t boss_idx = graph.kmer_to_boss_index(i);
+            if (dummy[boss_idx] || terminal[boss_idx]) {
+                succ_buf[r].push_back(0);
+            } else {
+                const BOSS::TAlphabet d = boss.get_W(boss_idx) % boss.alph_size;
+                uint64_t next = d ? graph.boss_to_kmer_index(boss.fwd(boss_idx, d)) : 0;
```

Comment:
Instead of called `fwd` on all non-dummy, non-terminal indices, won't it be faster to just call `rank(start, s)` for every character `s`, then keep updating a rank counter as you iterate through (since you already compute `get_W`)?

Reply:
So this way, the ranks will be precomputed and you'll only have to compute `select` for each iteration.

Reply:
One problem is that every time you hit a dummy or a terminal node, you'll have to resync the counts either with extra `get_W` or `rank` calls, which will slow it down a bit.

Reply:
It sounds like significantly more code to me, but I don't know how faster is this going to make the whole function. (We also call bwd, etc.)

Reply:
Also, this whole function is not critical to the application. If we can easily make it faster, we should do it, but if it's something more than straightforward, I don't it's worth it.